panic: Type checker panics on ref to object with composite keys

This was caught during the nightly fuzzing job. To reproduce:

```
$ docker run --rm -it openpolicyagent/opa:0.23.1
> x = {{}: {}}; x[1]
panic: unreachable

goroutine 1 [running]:
github.com/open-policy-agent/opa/types.TypeOf(0xbbf880, 0xc000487fb0, 0x1, 0x1)
        /go/src/github.com/open-policy-agent/opa/types/types.go:838 +0x5fc
github.com/open-policy-agent/opa/types.Keys(0xdef460, 0xc00000d160, 0xc00022a2c8, 0x409e8b)
        /go/src/github.com/open-policy-agent/opa/types/types.go:723 +0x131
github.com/open-policy-agent/opa/ast.(*refChecker).checkRefLeaf(0xc00013e810, 0xdef460, 0xc00000d160, 0xc000488e80, 0x2, 0x2, 0x1, 0xc00000d1a0)
        /go/src/github.com/open-policy-agent/opa/ast/check.go:648 +0x83
github.com/open-policy-agent/opa/ast.(*refChecker).checkRef(0xc00013e810, 0xc000489080, 0xc000487f20, 0xc000488e80, 0x2, 0x2, 0x0, 0x0)
        /go/src/github.com/open-policy-agent/opa/ast/check.go:591 +0x12a
github.com/open-policy-agent/opa/ast.(*refChecker).Visit(0xc00013e810, 0xc96600, 0xc00000cc00, 0x90)
        /go/src/github.com/open-policy-agent/opa/ast/check.go:548 +0x428
github.com/open-policy-agent/opa/ast.(*GenericVisitor).Walk(0xc00022a830, 0xc96600, 0xc00000cc00)
        /go/src/github.com/open-policy-agent/opa/ast/visit.go:270 +0x5b
github.com/open-policy-agent/opa/ast.(*GenericVisitor).Walk(0xc00022a830, 0xc6d700, 0xc00000cc20)
        /go/src/github.com/open-policy-agent/opa/ast/visit.go:332 +0xc0d
github.com/open-policy-agent/opa/ast.(*refChecker).Visit(0xc00013e810, 0xc9cbc0, 0xc00045c1e0, 0x203000)
        /go/src/github.com/open-policy-agent/opa/ast/check.go:540 +0x228
github.com/open-policy-agent/opa/ast.(*GenericVisitor).Walk(0xc00022aa68, 0xc9cbc0, 0xc00045c1e0)
        /go/src/github.com/open-policy-agent/opa/ast/visit.go:270 +0x5b
github.com/open-policy-agent/opa/ast.(*typeChecker).CheckBody.func1(0xc00045c1e0, 0x7f58172e3001)
```

--- a/ast/check.go
+++ b/ast/check.go
@@ -669,10 +669,6 @@ func (rc *refChecker) checkRefLeaf(tpe types.Type, ref Ref, idx int) *Error {
 		}
 
 	case *Array, Object, Set:
-		// Composite references operands may only be used with a set.
-		if !unifies(tpe, types.NewSet(types.A)) {
-			return newRefErrInvalid(ref[0].Location, rc.varRewriter(ref), idx, tpe, types.NewSet(types.A), nil)
-		}
 		if !unify1(rc.env, head, keys, false) {
 			return newRefErrInvalid(ref[0].Location, rc.varRewriter(ref), idx, rc.env.Get(head), keys, nil)
 		}

--- a/ast/check_test.go
+++ b/ast/check_test.go
@@ -110,6 +110,16 @@ func TestCheckInference(t *testing.T) {
 				nil,
 			),
 		}},
+		{"object-composite-ref-operand", `x = {{}: 1}; x[{}] = y`, map[Var]types.Type{
+			Var("x"): types.NewObject(
+				[]*types.StaticProperty{types.NewStaticProperty(
+					map[string]interface{}{},
+					types.N,
+				)},
+				nil,
+			),
+			Var("y"): types.N,
+		}},
 		{"sets", `x = {1, 2}; y = {{"foo", 1}, x}`, map[Var]types.Type{
 			Var("x"): types.NewSet(types.N),
 			Var("y"): types.NewSet(
@@ -795,12 +805,12 @@ func TestCheckRefErrInvalid(t *testing.T) {
 			oneOf: []Value{String("p"), String("q")},
 		},
 		{
-			note:  "composite ref into non-set",
+			note:  "composite ref operand",
 			query: `data.test.q[[1, 2]]`,
 			ref:   "data.test.q[[1, 2]]",
 			pos:   3,
-			have:  types.NewObject([]*types.StaticProperty{types.NewStaticProperty("bar", types.N), types.NewStaticProperty("foo", types.N)}, nil),
-			want:  types.NewSet(types.A),
+			have:  types.NewArray([]types.Type{types.N, types.N}, nil),
+			want:  types.S,
 		},
 		{
 			note:  "composite ref type error 1",
@@ -817,6 +827,14 @@ func TestCheckRefErrInvalid(t *testing.T) {
 			pos:   1,
 			have:  types.NewObject([]*types.StaticProperty{types.NewStaticProperty("a", types.S)}, nil),
 			want:  types.NewObject([]*types.StaticProperty{types.NewStaticProperty("a", types.N)}, nil),
+		},
+		{
+			note:  "composite ref type error 3 - array",
+			query: `a = [1,2,3]; a[{}] = b`,
+			ref:   `a[{}]`,
+			pos:   1,
+			have:  types.NewObject(nil, types.NewDynamicProperty(types.A, types.A)),
+			want:  types.N,
 		},
 	}
 

--- a/ast/compile.go
+++ b/ast/compile.go
@@ -3033,6 +3033,13 @@ func rewriteDynamicsComprehensionBody(original *Expr, f *equalityFactory, body B
 }
 
 func rewriteExprTermsInHead(gen *localVarGenerator, rule *Rule) {
+	for i := range rule.Head.Args {
+		support, output := expandExprTerm(gen, rule.Head.Args[i])
+		for j := range support {
+			rule.Body.Append(support[j])
+		}
+		rule.Head.Args[i] = output
+	}
 	if rule.Head.Key != nil {
 		support, output := expandExprTerm(gen, rule.Head.Key)
 		for i := range support {

--- a/ast/compile_test.go
+++ b/ast/compile_test.go
@@ -1197,6 +1197,17 @@ func TestCompilerRewriteExprTerms(t *testing.T) {
 			`,
 			expected: Errors{&Error{Message: "rule arguments cannot contain calls"}},
 		},
+		{
+			note: "indirect ref in args",
+			module: `
+				package test
+
+				f([1][0]) { true }`,
+			expected: `
+				package test
+
+				f(__local0__[0]) { true; __local0__ = [1] }`,
+		},
 	}
 
 	for _, tc := range cases {

--- a/types/types.go
+++ b/types/types.go
@@ -822,6 +822,15 @@ func TypeOf(x interface{}) Type {
 		return S
 	case json.Number:
 		return N
+	case map[string]interface{}:
+		// The ast.ValueToInterface() function returns ast.Object values as map[string]interface{}
+		// so map[string]interface{} must be handled here because the type checker uses the value
+		// to interface conversion when inferring object types.
+		static := make([]*StaticProperty, 0, len(x))
+		for k, v := range x {
+			static = append(static, NewStaticProperty(k, TypeOf(v)))
+		}
+		return NewObject(static, nil)
 	case map[interface{}]interface{}:
 		static := make([]*StaticProperty, 0, len(x))
 		for k, v := range x {

--- a/types/types_test.go
+++ b/types/types_test.go
@@ -298,6 +298,22 @@ func TestTypeOf(t *testing.T) {
 	}
 }
 
+func TestTypeOfMapOfString(t *testing.T) {
+	tpe := TypeOf(map[string]interface{}{
+		"foo": "bar",
+		"baz": "qux",
+	})
+
+	exp := NewObject([]*StaticProperty{
+		NewStaticProperty("foo", S),
+		NewStaticProperty("baz", S),
+	}, nil)
+
+	if Compare(exp, tpe) != 0 {
+		t.Fatalf("Expected %v but got: %v", exp, tpe)
+	}
+}
+
 func TestNil(t *testing.T) {
 
 	tpe := NewObject([]*StaticProperty{

